### ISSUE TITLE
docker login failed but webui works

### ISSUE THREAD

**Issue by sysreport** [goharbor/harbor#15289]
- harbor version: 2.3.0
- docker-compose version: docker-compose version 1.29.2, build 5becea4c
- docker version: 20.10.7
- OS: debian 10

Deployed with trivy and chartmuseum

`
docker-compose ps
      Name                     Command                  State                      Ports                
--------------------------------------------------------------------------------------------------------
chartmuseum         ./docker-entrypoint.sh           Up (healthy)                                       
harbor-core         /harbor/entrypoint.sh            Up (healthy)                                       
harbor-jobservice   /harbor/entrypoint.sh            Up (healthy)                                       
harbor-log          /bin/sh -c /usr/local/bin/ ...   Up (healthy)   127.0.0.1:1514->10514/tcp           
harbor-portal       nginx -g daemon off;             Up (healthy)                                       
nginx               nginx -g daemon off;             Up (healthy)   0.0.0.0:80->8080/tcp,:::80->8080/tcp
redis               redis-server /etc/redis.conf     Up (healthy)                                       
registry            /home/harbor/entrypoint.sh       Up (healthy)                                       
registryctl         /home/harbor/start.sh            Up (healthy)                                       
trivy-adapter       /home/scanner/entrypoint.sh      Up (healthy)
`

**Problem:**
WebUI is working fine and we can login to it. But a docker login command from docker cli is not working and comes back with unauthorized: authentication required

in the core.log i found only many of this lines:
`
Jul  8 10:46:05 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:05Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:08 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:08Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:08 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:08Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:08 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:08Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:08 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:08Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:18 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:18Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:18 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:18Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:18 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:18Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:18 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:18Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:54 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:54Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:54 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:54Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:55 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:55Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:55 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:55Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:58 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:58Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:58 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:58Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:58 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:58Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:46:58 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:46:58Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:47:00 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:47:00Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:47:00 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:47:00Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:47:00 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:47:00Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service: 
Jul  8 10:47:00 ip-192-168-112-1.eu-west-1.compute.internal core[3381]: 2021-07-08T10:47:00Z [ERROR] [/core/service/token/token.go:41]: Unable to handle service:
`

Special thanks for Help

**Comment by mlushpenko** [comment]
I wanted to open a new issue but it's probably the same. Docker login is failing, but after you login to the UI and try again, login works. Somehow those two things are related to each other. We have a whole team with the same problem, our docker commands are working only after we login to the web UI.

**Comment by phin1x** [comment]
Maybe the user needs to be onboarded first. Harbor creates every user in its internal database on its first login. It seams that the docker login webhook in the harbor core component doesn't onboard the user.

**Comment by ninjadq** [comment]
hi @sysreport , could you provide more logs with debug mode enabled

**Comment by mlushpenko** [comment]
@ninjadq I think because we have integration with Azure via OIDC, there is some conflict in the authentication flow:

```
/server/middleware/security/oidc_cli.go:62][requestID="8c238102913f9ae432053fa54e08f533"]: failed to verify secret, username: robot-k8s-pull, error: failed to get oidc user info, error: <QuerySeter> no row found  
```
`robot-k8s-pull` is a robot account and shouldn't go through any sort of OIDC check I think, but not sure

**Comment by wy65701436** [comment]
hi, please enable the debug log, and then you can get the debug log, like: `URL for token request: *****`

**Comment by Antiarchitect** [comment]
Seems I have similar issue. When using OIDC user email and CLI Secret (on Mac and Linux dev machines) there are cases when docker pull just stops working suddenly and only `docker login <servername>` saves the situation. Do not know if related.

**Comment by mlushpenko** [comment]
So, enabled debug logging and found a few things:

```
2021-07-16T14:32:09Z [DEBUG] [/controller/event/handler/webhook/artifact/artifact.go:77]: cannot find policy for PULL_ARTIFACT event: ID-1968, Repository-docker/test-api Tags-[2.0] Digest-sha256:2d2b4a1ace8025e14e57341cd309432e1ce38d16c84b55bc6f819c05810ec137 Operator-robot-k8s-pull OccurAt-2021-07-16 14:32:09
```

This one is perhaps not critical and I see a related issue https://github.com/goharbor/harbor/issues/13955
```
2021-07-16 17:37:24 | 2021-07-16T14:37:24Z [DEBUG] [/pkg/config/manager.go:139]: failed to get key oidc_user_claim, error: the configure value is not set
```

@wy65701436  And finally got proper log about CLI login issues, it felt like a refresh token issue and seems like it is indeed:
```
2021-07-19 16:23:08	2021-07-19T13:23:08Z [DEBUG] [/server/middleware/artifactinfo/artifact_info.go:52]: In artifact info middleware, url: /chartrepo/helm-elastic/charts/elasticsearch.tgz
2021-07-19 16:23:08	2021-07-19T13:23:08Z [DEBUG] [/pkg/oidc/secret.go:75]: Verifying the secret for user: mlushpenko
2021-07-19 16:23:08	2021-07-19T13:23:08Z [DEBUG] [/pkg/oidc/secret.go:104]: Refreshing token
2021-07-19 16:23:08	2021-07-19T13:23:08Z [ERROR] [/server/middleware/security/oidc_cli.go:62][requestID="860db526355d1d56de60d0eb019efced"]: failed to verify secret, username: maksym, error: failed to refresh token, username: maksym, error: oauth2: token expired and refresh token is not set
2021-07-19 16:23:08	2021-07-19T13:23:08Z [DEBUG] [/core/auth/authenticator.go:147]: Current AUTH_MODE is oidc_auth
2021-07-19 16:23:08	2021-07-19T13:23:08Z [ERROR] [/server/middleware/security/basic_auth.go:40][requestID="860db526355d1d56de60d0eb019efced"]: failed to authenticate maksym: not supported
2021-07-19 16:23:08	2021-07-19T13:23:08Z [DEBUG] [/server/middleware/security/unauthorized.go:28][requestID="e7d247062f6bf1811e6bad711ebb623a"]: an unauthorized security context generated for request GET /chartrepo/eiq-ops/charts/logging-3.0.0.tgz
2021-07-19 16:23:08	2021-07-19T13:23:08Z [DEBUG] [/server/middleware/security/unauthorized.go:28][requestID="860db526355d1d56de60d0eb019efced"]: an unauthorized security context generated for request GET /chartrepo/helm-elastic/charts/elasticsearch.tgz
2021-07-19 16:23:08	2021/07/19 13:23:08 Model:
2021-07-19 16:23:08	2021/07/19 13:23:08 r.r: sub, obj, act
2021-07-19 16:23:08	2021/07/19 13:23:08 p.p: sub, obj, act, eft
2021-07-19 16:23:08	2021/07/19 13:23:08 e.e: some(where (p_eft == allow)) && !some(where (p_eft == deny))
2021-07-19 16:23:08	2021/07/19 13:23:08 m.m: g(r_sub, p_sub) && keyMatch2(r_obj, p_obj) && (r_act == p_act || p_act == '*')
2021-07-19 16:23:08	2021/07/19 13:23:08 g.g: _, _
2021-07-19 16:23:08	2021/07/19 13:23:08 Policy:
2021-07-19 16:23:08	2021/07/19 13:23:08 p: sub, obj, act, eft: []
2021-07-19 16:23:08	2021/07/19 13:23:08 g: _, _: []
2021-07-19 16:23:08	2021/07/19 13:23:08 Role links for: g
2021-07-19 16:23:08	2021/07/19 13:23:08 
2021-07-19 16:23:08	2021/07/19 13:23:08 Request: anonymous, /project/607/helm-chart, read ---> false
2021-07-19 16:23:08	2021-07-19T13:23:08Z [DEBUG] [/lib/http/error.go:59]: {"errors":[{"code":"UNAUTHORIZED","message":"UnAuthorized"}]}
2021-07-19 16:23:08	2021/07/19 13:23:08 Model:
2021-07-19 16:23:08	2021/07/19 13:23:08 r.r: sub, obj, act
2021-07-19 16:23:08	2021/07/19 13:23:08 p.p: sub, obj, act, eft
2021-07-19 16:23:08	2021/07/19 13:23:08 e.e: some(where (p_eft == allow)) && !some(where (p_eft == deny))
2021-07-19 16:23:08	2021/07/19 13:23:08 m.m: g(r_sub, p_sub) && keyMatch2(r_obj, p_obj) && (r_act == p_act || p_act == '*')
2021-07-19 16:23:08	2021-07-19T13:23:08Z [DEBUG] [/lib/http/error.go:59]: {"errors":[{"code":"UNAUTHORIZED","message":"UnAuthorized"}]}
2021-07-19 16:23:08	2021/07/19 13:23:08 g.g: _, _
2021-07-19 16:23:08	2021/07/19 13:23:08 Policy:
2021-07-19 16:23:08	2021/07/19 13:23:08 p: sub, obj, act, eft: []
2021-07-19 16:23:08	2021/07/19 13:23:08 g: _, _: []
2021-07-19 16:23:08	2021/07/19 13:23:08 Role links for: g
2021-07-19 16:23:08	2021/07/19 13:23:08 
2021-07-19 16:23:08	2021/07/19 13:23:08 Request: anonymous, /project/3/helm-chart, read ---> false
```

**Comment by mlushpenko** [comment]
As far as I remember, it started working after adding `offline` claim to the token in admin settings
